crates/agent/src/reconcile/process.rs

```diff
@@ -0,0 +1,164 @@
+            })
+            .map_err(|e| {
+                error!("failed to start node process: {e:?}");
+                ReconcileError::SpawnError(e.to_string())
+            })
+    }
+
+    /// Returns true when the child process has not exited
+    pub fn is_running(&mut self) -> bool {
+        // This code is mutable because try_wait modifies the Child. Without
+        // mutability, the current running status would never be updated.
+        self.child.try_wait().is_ok_and(|status| status.is_none())
+    }
+
+    /// Check if the running binary matches the provided sha256 hash
+    pub fn is_sha256_eq(&self, sha256: &str) -> bool {
+        self.binary_sha256 == sha256
+    }
+
+    /// A helper function to gracefully shutdown the node process without
+    /// a reconciler
+    pub async fn graceful_shutdown(&mut self) {
+        if !self.is_running() {
+            return;
+        }
+
+        self.send_sigint();
+
+        select! {
+            _ = tokio::time::sleep(NODE_GRACEFUL_SHUTDOWN_TIMEOUT) => {
+                info!("Sending SIGKILL to node process");
+                self.send_sigkill();
+            },
+            _ = tokio::signal::ctrl_c() => {
+                info!("Received SIGINT, sending SIGKILL to node process");
+                self.send_sigkill();
+            },
+            _ = self.child.wait() => {
+                info!("Node process has exited gracefully");
+                return;
+            }
+        }
+
+        let _ = self.child.wait().await;
+        info!("Node process has exited");
+    }
+
+    /// Send a SIGINT to the child process
+    pub fn send_sigint(&mut self) -> bool {
+        use nix::{
+            sys::signal::{self, Signal},
+            unistd::Pid,
+        };
+
+        // prevent multiple sigints
+        if self.sigint_at.is_some() {
+            return false;
+        }
+
+        // obtain the id, or return false if the child is not running
+        let Some(id) = self.child.id() else {
+            return false;
+        };
+
+        // send SIGINT to the child process
+        signal::kill(Pid::from_raw(id as i32), Signal::SIGINT)
+            .inspect(|_| {
+                // update the sigint time if the sigint was successful
+                self.sigint_at = Some(Instant::now());
+            })
+            .is_ok()
+    }
+
+    /// Send a SIGKILL to the child process
+    pub fn send_sigkill(&mut self) -> bool {
+        // start_kill return Err if the process is already killed
+        self.child
+            .start_kill()
+            .inspect(|_| {
+                // update the kill time if the kill was successful
+                self.sigkill_at = Some(Instant::now());
+            })
+            .is_ok()
+    }
+}
+
+/// The EndProcessReconciler will return true when the child process has exited.
+/// It will wait NODE_GRACEFUL_SHUTDOWN_TIMEOUT seconds after sending a SIGINT
+/// before sending a SIGKILL (if the childi process has not exited),
```

Comment:
childi

Reply:
Gonna pretend this one is deliberate to thwart plagiarism without proper licensing rather than making a commit for one letter